README
NAME
    Text::ResusciAnneparser - Parser for XML logfiles of the Resusci Anne
    Skills Station software

VERSION
    version 0.02

SYNOPSIS
    my $certificates = Text::ResusciAnneparser->new(infile =>
    'certificates.xml');

DESCRIPTION
    The Resusci Anne Skills Station is a basic life support training station
    used by people involved in first-line support in healthcare. The
    training station keeps track of who trained when. This module enables
    parsing the xml output file to be able to process the data.

METHODS
  "new(%parameters)"
    This constructor returns a new Text::ResusciAnneparser object. Supported
    parameters are listed below

    infile
        The input file containing the raw data log of the skill station
        software

  "certified"
    Returns a hash of people who received a valid training certificate. The
    hash contains keys with the training dates in the format YYYY-MM-DD. The
    value attached to a date key in the hash is an array of people.

    A single person entry is a hash containing the givenname and the
    familiname of a person.

    E.g. '2013-04-07' => [ { 'givenname' => 'Piet', 'familyname' => 'Konijn'
    } ], '2013-03-25' => [ { 'givenname' => 'Zjuul', 'familyname' => 'Cesar'
    }, { 'givenname' => 'Pette', 'familyname' => 'Sjiekke' } ]

  "in_training"
    Returns an array of people who started the exercise but who did not
    completed it and hence have not received a certificate yet

  BUILD
    Helper function to run custome code after the object has been created by
    Moose.

AUTHOR
    Lieven Hollevoet <hollie@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Lieven Hollevoet.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2012-06-21 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.02 at 2013-06-21 18:57:10 +0000
-----------------------------------------

  Change: c4424e28d520f389c07e3c5c870680e5e07862df
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-21 20:55:54 +0000

    Bumped version to 0.02 and fixed release tests 

  Change: dea87c91f0659e80e24d22bb69e7a647793fe0a1
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-21 20:52:34 +0000

    Added removal of leading/trailing spaces in the name fields 

  Change: eb3fddf853eae6d079bac1815654bb5d1acdd4c2
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-17 20:17:15 +0000

    Removed overly strict Perl version requirement 

  Change: 800f0c269a9494cb1c218e298b5114d99d76c397
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 23:00:48 +0000

    v0.01 

  Change: 9feebd1936660e6f4bf8fc34e2533e64934486d0
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 22:59:13 +0000

    Ignore release folders in git 

  Change: 91e84c90734ac8bfa4208afac45aba8032220c85
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 22:55:05 +0000

    Fixed release tests reported by Dist::Zilla 

  Change: 0d09da3982a21e3f19a4cc5cf0ef8a22e273ee55
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 22:41:12 +0000

    Basic verification OK for parsing the output xml file 

  Change: 45599a0a50772051dd33c19fa3bad5145a2a92d4
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 14:10:53 +0000

    Merge remote-tracking branch 'origin/master' 

  Change: b154080f52f9706b58aac88447f53c245c85ccee
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 14:10:23 +0000

    Added dist.ini file as a start 

  Change: aa1b01ce14640e8a8d9770534b528fb9c7e24b75
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 05:09:36 +0000

    Initial commit 

-----------------------------------------
version 0.01 at 2013-06-15 21:00:49 +0000
-----------------------------------------

  Change: 79cb043de02677eea8988ea59cee13d522e12308
  Author: Lieven Hollevoet <hollie@lika.be>
  Date : 2013-06-15 23:00:49 +0000

    Build results of 800f0c2 (on master) 

================
End of releases.
================


Makefile.PL

use strict;
use warnings;

use 5.012;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Parser for XML logfiles of the Resusci Anne Skills Station software",
  "AUTHOR" => "Lieven Hollevoet <hollie\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Text-ResusciAnneparser",
  "EXE_FILES" => [
    "bin/basic.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "Text::ResusciAnneparser",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "DateTime" => 0,
    "Moose" => 0,
    "XML::Simple" => 0,
    "autodie" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0
  },
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Text/ResusciAnneparser.pm
use strict;    # To keep Test::Perl::Critic happy, Moose does enable this too...

package Text::ResusciAnneparser;
{
  $Text::ResusciAnneparser::VERSION = '0.02';
}

use Moose;
use namespace::autoclean;
use 5.012;
use autodie;

use DateTime;
use XML::Simple qw(:strict);
use Data::Dumper;

has infile => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

use Carp qw/croak carp/;

# Ensure we read the inputfile after constructing the object
sub BUILD {
    my $self = shift;
    $self->{_data} = {};
    $self->_read_infile;
}

sub _read_infile {

    my $self = shift;

    my $certificates =
      XMLin( $self->{infile}, ForceArray => 1, KeyAttr => { user => 'login' } );

# Sort users according to the ones who got a certificate and the ones who did not
    foreach my $user ( keys %{ $certificates->{user} } ) {

        my $fname = $certificates->{user}->{$user}->{familyname};
        my $gname = $certificates->{user}->{$user}->{givenname};

        # Ensure no leading/trailing spaces are in the name
        $fname =~ s/^\s+//; # strip white space from the beginning
        $fname =~ s/\s+$//; # strip white space from the end
        $gname =~ s/^\s+//; # strip white space from the beginning
        $gname =~ s/\s+$//; # strip white space from the end

        my $names = {
            'givenname'  => $gname,
            'familyname' => $fname
        };

        if ( defined $certificates->{user}->{$user}->{'course'} ) {
            my $course = $certificates->{user}->{$user}->{'course'}->[0];
            my $dt     = DateTime->new(
                year  => $course->{year},
                month => $course->{month},
                day   => $course->{day}
            );

            # Make an entry under {certs}
            # Entry contains the course date and email address
            push( @{ $self->{_data}->{certs}->{ $dt->ymd } }, $names );
        } else {
            push( @{ $self->{_data}->{training} }, $names );
        }
    }

}

sub certified {
    my $self = shift;
    return $self->{_data}->{certs};
}

sub in_training {
    my $self = shift;
    return $self->{_data}->{training};
}

# Speed up the Moose object construction
__PACKAGE__->meta->make_immutable;
no Moose;
1;

# ABSTRACT: Parser for XML logfiles of the Resusci Anne Skills Station software

__END__

=pod

=head1 NAME

Text::ResusciAnneparser - Parser for XML logfiles of the Resusci Anne Skills Station software

=head1 VERSION

version 0.02

=head1 SYNOPSIS

my $certificates = Text::ResusciAnneparser->new(infile => 'certificates.xml');

=head1 DESCRIPTION

The Resusci Anne Skills Station is a basic life support training station used by people
involved in first-line support in healthcare.
The training station keeps track of who trained when. This module enables parsing the
xml output file to be able to process the data.

=head1 METHODS

=head2 C<new(%parameters)>

This constructor returns a new Text::ResusciAnneparser object. Supported parameters are listed below

=over

=item infile

The input file containing the raw data log of the skill station software

=back

=head2 C<certified>

Returns a hash of people who received a valid training certificate. The hash contains keys with the
training dates in the format YYYY-MM-DD. The value attached to a date key in the hash is an array
of people.

A single person entry is a hash containing the givenname and the familiname of a person.

E.g.
           '2013-04-07' => [
                             {
                               'givenname' => 'Piet',
                               'familyname' => 'Konijn'
                             }
                            ],
           '2013-03-25' => [
                             {
                               'givenname' => 'Zjuul',
                               'familyname' => 'Cesar'
                             },
                             {
                               'givenname' => 'Pette',
                               'familyname' => 'Sjiekke'
                             }
                           ]

=head2 C<in_training>

Returns an array of people who started the exercise but who did not completed it and hence have not received
a certificate yet

=head2 BUILD

Helper function to run custome code after the object has been created by Moose.

=head1 AUTHOR

Lieven Hollevoet <hollie@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Lieven Hollevoet.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
